nibabel/funcs.py

```diff
@@ -88,43 +88,81 @@ def squeeze_image(img):
-    ''' Concatenate images in list to single image, along last dimension
+def _shape_equal_excluding(shape1, shape2, exclude_axes):
+    """ Helper function to compare two array shapes, excluding any
+    axis specified."""
+
+    if len(shape1) != len(shape2):
+        return False
+
+    idx_mask = np.ones((len(shape1),), dtype=bool)
+    idx_mask[exclude_axes] = False
+    return np.array_equal(np.asarray(shape1)[idx_mask],
+                          np.asarray(shape2)[idx_mask])
+
+
+def concat_images(images, check_affines=True, axis=None):
+    ''' Concatenate images in list to single image, along specified dimension
 
     Parameters
     ----------
     images : sequence
-       sequence of ``SpatialImage`` or of filenames\s
+       sequence of ``SpatialImage`` or filenames of the same dimensionality\s
     check_affines : {True, False}, optional
        If True, then check that all the affines for `images` are nearly
        the same, raising a ``ValueError`` otherwise.  Default is True
-
+    axis : None or int, optional
+        If None, concatenates on a new dimension.  This rrequires all images
+           to be the same shape).
+        If not None, concatenates on the specified dimension.  This requires
+           all images to be the same shape, except on the specified dimension.
+           For 4D images, axis must be between -2 and 3.
     Returns
     -------
     concat_img : ``SpatialImage``
        New image resulting from concatenating `images` across last
        dimension
     '''
+
     n_imgs = len(images)
-    img0 = images[0]
-    is_filename = False
-    if not hasattr(img0, 'get_data'):
-        img0 = load(img0)
-        is_filename = True
-    i0shape = img0.shape
-    affine = img0.affine
-    header = img0.header
-    out_shape = (n_imgs, ) + i0shape
-    out_data = np.empty(out_shape)
+    if n_imgs == 0:
+        raise ValueError("Cannot concatenate an empty list of images.")
+
     for i, img in enumerate(images):
-        if is_filename:
+        if not hasattr(img, 'get_data'):
             img = load(img)
-        if check_affines:
-            if not np.all(img.affine == affine):
-                raise ValueError('Affines do not match')
+
+        if i == 0:  # first image, initialize data from loaded image
+            affine = img.affine
+            header = img.header
+            shape = img.shape
+            klass = img.__class__
+
+            if axis is None:  # collect images in output array for efficiency
+                out_shape = (n_imgs, ) + shape
+                out_data = np.empty(out_shape)
+            else:  # collect images in list for use with np.concatenate
+                out_data = [None] * n_imgs
+
+        elif check_affines and not np.all(img.affine == affine):
+            raise ValueError('Affines do not match')
+
+        elif ((axis is None and not np.array_equal(shape, img.shape)) or
+              (axis is not None and not _shape_equal_excluding(shape, img.shape,
```

Comment:
It will error, but the error message is pretty cryptic.  I added this logic in after seeing the error.